src/backend/puppet/mod.rs

```diff
@@ -0,0 +1,165 @@
+//! Puppet backend
+use std::thread;
+
+use std::collections::LinkedList;
+use std::time::{Duration, Instant};
+use std::cell::{Cell, RefCell};
+
+use crossbeam_channel::{self, Receiver, Sender};
+
+use backend;
+use event::Event;
+use theme;
+use vec::Vec2;
+use enumset::EnumSet;
+
+use theme::Style;
+
+pub mod observed;
+use self::observed::ObservedFrame;
+use std::rc::Rc;
+use unicode_segmentation::UnicodeSegmentation;
+use backend::puppet::observed::ObservedChar;
+use theme::Effect;
+use theme::ColorPair;
+use std::cell::Ref;
+
+///
+pub struct Backend {
+    size : Vec2,
+    inner_sender: Sender<Option<Event>>,
+    inner_receiver: Receiver<Option<Event>>,
+    frames : LinkedList<ObservedFrame>,
+    current_frame : RefCell<ObservedFrame>,
+    current_color : RefCell<Rc<theme::ColorPair>>,
+    current_effect : RefCell<Rc<EnumSet<theme::Effect>>>,
+    current_cursor : Vec2,
+}
+
+impl Backend {
+    ///
+    pub fn init() -> Box<backend::Backend>
+        where
+            Self: Sized,
+    {
+        let (inner_sender, inner_receiver) = crossbeam_channel::bounded(1);
+        let size = Vec2::new(100, 100); //TODO something less random
+        let color_pair = theme::ColorPair {
+            front: theme::Color::TerminalDefault,
+            back : theme::Color::Dark(theme::BaseColor::Black)
+        };
+
+        Box::new(Backend {
+            size,
+            inner_sender,
+            inner_receiver,
+            frames : LinkedList::new(),
+            current_frame : RefCell::new(ObservedFrame::new(size)),
+            current_color : RefCell::new(Rc::new(color_pair)),
+            current_effect : RefCell::new(Rc::new(EnumSet::new())),
+            current_cursor: Vec2::new(0,0)
+        })
+    }
+
+    ///
+    pub fn current_frame(&self) -> Ref<ObservedFrame> {
+        self.current_frame.borrow()
+    }
+
+    ///
+    pub fn last_frame(&self) -> Option<&ObservedFrame> {
+        self.frames.front()
+    }
+}
+
+impl backend::Backend for Backend {
+    fn finish(&mut self) {
```

Comment:
`finish()` is called at the end, when the backend is about to be dropped and we want to "clean up" the terminal.
It's not called after each frame, you may instead be interested in `refresh()`.